### ISSUE TITLE
Dependencies not injected on classes without @inject using Python 3

### ISSUE THREAD

**Issue by asyncee** [alecthomas/injector#102]
Hello, documentation states following:

If you use Python 3 you can make Injector use constructor parameter annotations to determine class dependencies. The following code:

 ```
class B:
    @inject
    def __init__(self, a: A):
        self.a = a
```
can now be written as:

```
class B:
    def __init__(self, a:A):
        self.a = a
```

I can not get it working:

```
from injector import *


class A:
    pass


class B:
    def __init__(self, a: A):
        self.a = a


class MyModule(Module):
    @singleton
    @provider
    def provide_a(self) -> A:
        return A()


injector = Injector(MyModule)
injector.get(B)
```

If `@inject` is added to the B's constructor, then everything working fine.

**Comment by asyncee** [comment]
The exception is:

```
TypeError: __init__() missing 1 required positional argument: 'a'

During handling of the above exception, another exception occurred:

injector.CallError: Call to B.__init__() failed: __init__() missing 1 required positional argument: 'a' (injection stack: [])
```

**Comment by jstasiak** [comment]
Hmm, the documentation is outdated, not sure why (https://readthedocs.org/projects/injector/builds/ says last build is 1 year 4 months old). @alecthomas help?

This piece of documentation has been removed in 5e435c00233be8d54e8dbf6ba17a699ca52528b5 as it wasn't even correct, `@inject` is required.

**Comment by asyncee** [comment]
Could you clarify reasons why is this feature was dropped? It looks so convenient to me because of:

1. simplicity of use
2. one does not need to "pollute" business code (possibly pure python without dependencies) with implementation details like injector 

Thanks!

**Comment by jstasiak** [comment]
Readability and documenting intent (if you see `@inject` you expect for at least some of the arguments to be provided by the dependency injection mechanism, not by the client code).

If you don't mind me asking - what's the situation that results in "pure python without dependencies" and Injector-provided arguments at the same time?

**Comment by asyncee** [comment]
Sure. For example, i have a library that is written with pure python and have no third-party dependencies.

This library is packaged and used in several projects.

For example, there are a class `A` that has some public dependencies from the same library. Every client instantiates this class manually with providing an instantiated dependencies to it.

So the library is third-party dependency for other projects itself and can not depend on injector.  But the projects can use injector. So to build the `A` class one need to write a module with a factory function that creates an instance and then call `injector.get(A)` instead of just calling `injector.get(A)`.

I hope i was accurate enough to describe the situation. Also, it is okay for me for keeping "@inject" required, because explicit is better than implicit.

**Comment by jstasiak** [comment]
Yes, thank you, that was informative. I'd like to add one thing - decorating constructors with `@inject` doesn't make your library depend on Injector strictly speaking - yes, you need to be able to import `inject` and the `injector` package needs to be installed but the client code is free to keep constructing the objects manually. `@inject` doesn't prohibit manual instantiation.

**Comment by rustybox** [comment]
it's a shame the decorator is required, you could presumably figure out the positional args at the point binder.bind is called.

I'd much prefer to register all my services with binder.bind instead of making modules take a dependency for the decorator.

**Comment by jstasiak** [comment]
The presence of the decorator is, at this point, a design decision meant to document that dependency injection is expected to take place, it's not intrinsically required and serves (see the implementation) as just a marker really. So whether it's a shame or not depends on a point of view.

**Comment by rustybox** [comment]
when you say it's not required, it's not obvious how to get an Injector instance to be able to resolve even with explicit calls to binder.bind

```python
from injector import Injector, ClassProvider

class Inner:
    def __init__(self):
        self.forty_two = 42

class Outer:
    def __init__(self, inner: Inner):
        self.inner = inner

def configure(binder):
    binder.bind(Outer, to=ClassProvider(Outer))
    binder.bind(Inner, to=ClassProvider(Inner))

injector = Injector(configure)

inner = injector.get(Inner)
print(inner.forty_two)

# Throws
outer = injector.get(Outer)
print(outer.inner.forty_two)
```

**Comment by kwvanderlinde** [comment]
I've just started trying out `injector` on a new project and it looks quite interesting. But I would have to agree that requiring `@inject` is an unfortunate choice. Although it's a common choice for DI frameworks to make, IMO it is not a good choice. It's actually one of the first things I look at when I consider adopting a DI framework.

Here's some theoretical reasons why it doesn't make sense to require `@inject`:
1. "Dependency injection" is a general technique and does not imply the use of a DI framework. A framework can be used, but we can also create object graphs "by hand" - the two are functionally equivalent. As such, classes shouldn't need to broadcast their availability to DI frameworks.
2. From the perspective of the class using `@inject`, it makes no difference whether it is being created by a container or some other way. As long as the dependencies are provided, it shouldn't care how it got them. In other words, `@inject` is not something the class cares about, it's something the application cares about. The responsibilities here are all wrong.
3. One reason for other DI frameworks to require the equivalent of `@inject` is that the some other languages permit overloaded constructors. In that case `@inject` can serve to disambiguate which constructor to use. However, python does not have this issue.

There's also one really important practical reason to consider: it can actually hinder adoption of this library. In the past, I've had to search for DI frameworks for large projects. But we passed on a lot of - otherwise good - frameworks because they forced us to either add `@inject` to a lot of code, or to hand-write a bunch of providers, or both (we can't just modify 3rd party libraries). On top of that, if the annotations were non-standard, we would be effectively locked into using that specific framework. It can be hard to justify such things, despite any benefits we might derive. But when the framework just allows us to use a top-level injector without all the extra work, then it becomes a really attractive choice.

**Comment by asyncee** [comment]
I should note that we are using injector in few large projects and many smaller ones (microservices). We are using DDD approach, so injector and it's `@inject` decorator is just a cross-cutting concern, just like logging, etc. With such approach `@inject` becomes a no problem from theoretical point of view.

It could be better to not to use that DSL to annotate constructor injection, but anyway it is working perfectly for us. Also, we are not defining or using binder for each of dependency, but just for specific implementations of abstract classes (interfaces) that can not be chosen automatically and provided from nowhere. The code stays clean and simple.

**Comment by jstasiak** [comment]
@kwvanderlinde In case I wasn't clear in the earlier comments, `@inject`'s presence (together with `@noninjectable` etc.) is to document availability/expectation of DI using a DI framework to the *programmers*, not to the DI framework. The framework could work just fine without it. As such, while your points may apply in principle, they just represent one set of values (and design decisions) while others exist. I hope that you can see that in this light and in this context "hindering adoption of this library" is a non-argument.

That said, adding an opt-in way to assume `@inject` in every situation it's expected now would not be out of question I think.

**Comment by rustybox** [comment]
I don’t think announcing that an ioc container is expected to be used is useful, the classes constructor announces it needs dependencies that is enough, a readme file cabinet explain to new comers that an ioc container is used.

The decorator is enough to prevent me making use of this project as it would essentially be required for every application service or infrastructure class I ever wrote, to at some point replace the ioc container would therefore require making changes to every single class in the project whereas with a more imperative style I would only have to go to a single place (or a couple of modules) to update the configuration.

I hope you don’t opt for assuming @inject for all constructors I think that would be a mistake, I just wish there was both an imperative and declarative way to configure based on the needs of a given project.

**Comment by davidparsson** [comment]
@netijoe, the logic for injecting dependencies is currently implemented in the `@inject` decorator. Does not that count as a technical reason?

**Comment by netijoe** [comment]
Thank you for the response. I based my post on this one: 

> The presence of the decorator is, at this point, a design decision meant to document that dependency injection is expected to take place, it's not intrinsically required and serves (see the implementation) as just a marker really. So whether it's a shame or not depends on a point of view.

It seemed like it should work without `@inject` when `binder.bind` is used.
